ethcore/evm/src/interpreter/mod.rs

```diff
@@ -102,127 +103,234 @@ enum InstructionResult<Gas> {
+	/// Input data.
+	pub data: Option<Bytes>,
+	/// Type of call
+	pub call_type: CallType,
+	/// Param types encoding
+	pub params_type: ParamsType,
+}
+
+impl From<ActionParams> for InterpreterParams {
+	fn from(params: ActionParams) -> Self {
+		InterpreterParams {
+			code_address: params.code_address,
+			code_hash: params.code_hash,
+			address: params.address,
+			sender: params.sender,
+			origin: params.origin,
+			gas: params.gas,
+			gas_price: params.gas_price,
+			value: params.value,
+			data: params.data,
+			call_type: params.call_type,
+			params_type: params.params_type,
+		}
+	}
+}
+
+/// Stepping result returned by interpreter.
+pub enum InterpreterResult {
+	/// The VM has already stopped.
+	Stopped,
+	/// The VM has just finished execution in the current step.
+	Done(vm::Result<GasLeft>),
+	/// The VM can continue to run.
+	Continue,
+}
+
 /// Intepreter EVM implementation
 pub struct Interpreter<Cost: CostType> {
 	mem: Vec<u8>,
 	cache: Arc<SharedCache>,
+	params: InterpreterParams,
+	reader: CodeReader,
 	return_data: ReturnData,
+	informant: informant::EvmInformant,
+	do_trace: bool,
+	done: bool,
+	valid_jump_destinations: Option<Arc<BitSet>>,
+	gasometer: Gasometer<Cost>,
+	stack: VecStack<U256>,
 	_type: PhantomData<Cost>,
 }
 
 impl<Cost: CostType> vm::Vm for Interpreter<Cost> {
-	fn exec(&mut self, params: ActionParams, ext: &mut vm::Ext) -> vm::Result<GasLeft> {
-		self.mem.clear();
+	fn exec(&mut self, ext: &mut vm::Ext) -> vm::Result<GasLeft> {
+		loop {
+			let result = self.step(ext);
+			match result {
+				InterpreterResult::Continue => {},
+				InterpreterResult::Done(value) => return value,
+				InterpreterResult::Stopped => panic!("Attempted to execute an already stopped VM.")
+			}
+		}
+	}
+}
 
-		let mut informant = informant::EvmInformant::new(ext.depth());
-		let mut do_trace = true;
+impl<Cost: CostType> Interpreter<Cost> {
+	/// Create a new `Interpreter` instance with shared cache.
+	pub fn new(mut params: ActionParams, cache: Arc<SharedCache>, ext: &vm::Ext) -> vm::Result<Interpreter<Cost>> {
+		let reader = CodeReader::new(params.code.take().expect("VM always called with code; qed"));
+		let params = InterpreterParams::from(params);
+		let informant = informant::EvmInformant::new(ext.depth());
+		let valid_jump_destinations = None;
+		let gasometer = Gasometer::<Cost>::new(Cost::from_u256(params.gas)?);
+		let stack = VecStack::with_capacity(ext.schedule().stack_limit, U256::zero());
+
+		Ok(Interpreter {
+			cache, params, reader, informant,
+			valid_jump_destinations, gasometer, stack,
+			done: false,
+			do_trace: true,
+			mem: Vec::new(),
+			return_data: ReturnData::empty(),
+			_type: PhantomData,
+		})
+	}
 
-		let code = &params.code.as_ref().expect("exec always called with code; qed");
-		let mut valid_jump_destinations = None;
+	/// Execute a single step on the VM.
+	#[inline(always)]
+	pub fn step(&mut self, ext: &mut vm::Ext) -> InterpreterResult {
+		macro_rules! try_or_done {
```

Comment:
Is it really worth to have a macro for that?
Maybe sth like this:
```
#[inline(always)]
fn try_or_done<F>(x: Result<F, vm::Error>) -> Result<F, InterpreterResult> {
 x.map_err(|e| InterpreterResult::Done(Err(e))
}

// and used like this:
try_or_done(self.verify_instruction(ext, instruction, info))?;
```

Or maybe even add `From<vm::Error> for InterpreterResult` and convert errors automatically via `?`?

Reply:
The issue is that `?` only works for `Result` like things with `Ok` and `Err`. So we can't use any `?` syntax in `step` right now, unless we change `InterpreterResult` to be a `Result<A, B>`. This indeed can be done via:

* Return `Ok(result)`, which is the same case for `InterpreterResult::Done(_)`.
* Return error case for `InterpreterResult::Stopped` and `InterpreterResult::Continue`.

However, I think it may be a little bit confusing as `Ok`/`Err` do not really apply here.

Reply:
Oh, that's true. Forgot that it only works for `Result` on the return type as well, anyway might be worth extracting the stuff in `inner` to a separate function and maybe use a helper function instead of a macro. What do you think?

Reply:
I think I found a way: we can return `Result<(), InterpreterResult>` as the inner function and always return `Err`. The compiler would happily accept `?` syntax with that. And we would just need to make sure the inner function never returns `Ok(())`.

This detail is wrapped in `step_inner` so it doesn't affect public interface.